# Project 3 - Data Visualization of Ford Go-Bike System
## by (Ebenezer Acquah)

Table of Content

- [Introduction](#intro)
-[Prelimary Investigation](#investigation)
- [Data Wrangling](#wrangling)
- [Feature Engineering](#engineering)
- [Univariate Exploration](#univariate)
- [Bivariate Exploration](#bivariate)
- [Multivariate Exploration](#multivariate)
- [Summary](#summary)

### <a id='intro'> Introduction</a>

> This dataset provides information about individual rides made in a bike-sharing system in the San Francisco Bay Area. The dataset has 183412 rows and 16 columns

### <a id='investigation'>Prelimary Investigation</a>

Import Libraries

In [1]:
# import all packages and set plots to be embedded inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb

%matplotlib inline

# suppress warnings from final output
import warnings
warnings.simplefilter("ignore")

In [2]:
# load in the dataset into a pandas dataframe

# load
df = pd.read_csv("/home/eben-success/Desktop/Local/Udacity/Udacity/Udacity_Data_Analysis_Projects/Project 3/201902-fordgobike-tripdata.csv")


In [3]:
df_copy = df.copy()

In [4]:
df.head()

duration_sec                start_time                  end_time  \
0         52185  2019-02-28 17:32:10.1450  2019-03-01 08:01:55.9750   
1         42521  2019-02-28 18:53:21.7890  2019-03-01 06:42:03.0560   
2         61854  2019-02-28 12:13:13.2180  2019-03-01 05:24:08.1460   
3         36490  2019-02-28 17:54:26.0100  2019-03-01 04:02:36.8420   
4          1585  2019-02-28 23:54:18.5490  2019-03-01 00:20:44.0740   

   start_station_id                                start_station_name  \
0              21.0  Montgomery St BART Station (Market St at 2nd St)   
1              23.0                     The Embarcadero at Steuart St   
2              86.0                           Market St at Dolores St   
3             375.0                           Grove St at Masonic Ave   
4               7.0                               Frank H Ogawa Plaza   

   start_station_latitude  start_station_longitude  end_station_id  \
0               37.789625              -122.400811            13.0   
1               37.791464              -122.391034            81.0   
2               37.769305              -122.426826             3.0   
3               37.774836              -122.446546            70.0   
4               37.804562              -122.271738           222.0   

                               end_station_name  end_station_latitude  \
0                Commercial St at Montgomery St             37.794231   
1                            Berry St at 4th St             37.775880   
2  Powell St BART Station (Market St at 4th St)             37.786375   
3                        Central Ave at Fell St             37.773311   
4                         10th Ave at E 15th St             37.792714   

   end_station_longitude  bike_id   user_type  member_birth_year  \
0            -122.402923     4902    Customer             1984.0   
1            -122.393170     2535    Customer                NaN   
2            -122.404904     5905    Customer             1972.0   
3            -122.444293     6638  Subscriber             1989.0   
4            -122.248780     4898  Subscriber             1974.0   

  member_gender bike_share_for_all_trip  
0          Male                      No  
1           NaN                      No  
2          Male                      No  
3         Other                      No  
4          Male                     Yes

In [5]:
# check the number of rows and columns in the dataset
df.shape

(183412, 16)

In [6]:
# check tthe unique values in the target datasets (bike_share_for_all_trip)
df.bike_share_for_all_trip.unique()

array(['No', 'Yes'], dtype=object)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 183412 entries, 0 to 183411
Data columns (total 16 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   duration_sec             183412 non-null  int64  
 1   start_time               183412 non-null  object 
 2   end_time                 183412 non-null  object 
 3   start_station_id         183215 non-null  float64
 4   start_station_name       183215 non-null  object 
 5   start_station_latitude   183412 non-null  float64
 6   start_station_longitude  183412 non-null  float64
 7   end_station_id           183215 non-null  float64
 8   end_station_name         183215 non-null  object 
 9   end_station_latitude     183412 non-null  float64
 10  end_station_longitude    183412 non-null  float64
 11  bike_id                  183412 non-null  int64  
 12  user_type                183412 non-null  object 
 13  member_birth_year        175147 non-null  float64
 14  memb

In [8]:
df.isnull().sum()

duration_sec                  0
start_time                    0
end_time                      0
start_station_id            197
start_station_name          197
start_station_latitude        0
start_station_longitude       0
end_station_id              197
end_station_name            197
end_station_latitude          0
end_station_longitude         0
bike_id                       0
user_type                     0
member_birth_year          8265
member_gender              8265
bike_share_for_all_trip       0
dtype: int64

From the summary above, there are 197 missing values in start_station_id, start_station_name, end_station_id, end_station_name and 8265 missing values in member_birth_year and member_gender.

In [9]:
df.duplicated().sum ()

0

## <a id ='wrangling'>Data Wrangling</a>

Checking for missing values 

In [10]:
# Find the percentage of missing values

(df.isna().sum() / df.shape[0]) * 100

duration_sec               0.000000
start_time                 0.000000
end_time                   0.000000
start_station_id           0.107408
start_station_name         0.107408
start_station_latitude     0.000000
start_station_longitude    0.000000
end_station_id             0.107408
end_station_name           0.107408
end_station_latitude       0.000000
end_station_longitude      0.000000
bike_id                    0.000000
user_type                  0.000000
member_birth_year          4.506248
member_gender              4.506248
bike_share_for_all_trip    0.000000
dtype: float64

Fill missing values with the mode

In [11]:
for col in ['start_station_id', 'end_station_id', 'member_birth_year', 'member_gender' , 'end_station_name', 'start_station_name']:
    df[col] = df[col].fillna(df[col].mode()[0])
    
df['member_gender'] = df['member_gender'].fillna("Male")

#### Investigations made so far

>- Most of the entries have incorrect datatypes
>- Some columns have missing data

## <a id='engineering'>Feature Engineering</a>

Correcting Datatypes

In [12]:
# convert start_time and end_time to datetime format
df.start_time = pd.to_datetime(df. start_time)
df.end_time = pd.to_datetime(df.end_time)

# convert the start_station_id and end_stattion_id to strings
df.start_station_id = df.start_station_id.astype(str)

df.end_station_id = df.end_station_id.astype(str)

# convert the bike_id to string
df.bike_id = df.bike_id.astype(str)

# convert the user type and member_gender to categories
df.user_type = df.user_type.astype('category')

df.member_gender = df.member_gender.astype('category')

In [13]:
#checking to see if changes have been implemented
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 183412 entries, 0 to 183411
Data columns (total 16 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   duration_sec             183412 non-null  int64         
 1   start_time               183412 non-null  datetime64[ns]
 2   end_time                 183412 non-null  datetime64[ns]
 3   start_station_id         183412 non-null  object        
 4   start_station_name       183412 non-null  object        
 5   start_station_latitude   183412 non-null  float64       
 6   start_station_longitude  183412 non-null  float64       
 7   end_station_id           183412 non-null  object        
 8   end_station_name         183412 non-null  object        
 9   end_station_latitude     183412 non-null  float64       
 10  end_station_longitude    183412 non-null  float64       
 11  bike_id                  183412 non-null  object        
 12  user_type       

Fill the missing values with the mode

Recheck the percentage of missing values 

In [14]:
(df.isna().sum() / df.shape[0]) * 100

duration_sec               0.0
start_time                 0.0
end_time                   0.0
start_station_id           0.0
start_station_name         0.0
start_station_latitude     0.0
start_station_longitude    0.0
end_station_id             0.0
end_station_name           0.0
end_station_latitude       0.0
end_station_longitude      0.0
bike_id                    0.0
user_type                  0.0
member_birth_year          0.0
member_gender              0.0
bike_share_for_all_trip    0.0
dtype: float64

I will convert the duration_secs in seconds to duration_mins in minutes so that it would look more friendly. 

In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 183412 entries, 0 to 183411
Data columns (total 16 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   start_time               183412 non-null  datetime64[ns]
 1   end_time                 183412 non-null  datetime64[ns]
 2   start_station_id         183412 non-null  object        
 3   start_station_name       183412 non-null  object        
 4   start_station_latitude   183412 non-null  float64       
 5   start_station_longitude  183412 non-null  float64       
 6   end_station_id           183412 non-null  object        
 7   end_station_name         183412 non-null  object        
 8   end_station_latitude     183412 non-null  float64       
 9   end_station_longitude    183412 non-null  float64       
 10  bike_id                  183412 non-null  object        
 11  user_type                183412 non-null  category      
 12  member_birth_yea

In [32]:
df['duration_min'] = (df['duration_sec'] / 60)
df['duration_min'] = df['duration_min'].astype(int)

# Dropping the duration_sec columns

df.drop(['duration_sec'], axis=1, inplace=True)


Since the member_birth_year is given, I would calculate the member's present age

In [35]:
df['member_age'] = 2021 - df['member_birth_year']

# converting member age from float to int

df['member_age'] = df['member_age'].astype(int)

df.drop(columns=['member_birth_year'], inplace=True)

In [36]:
df['member_age'].describe()

count    183412.000000
mean         36.049653
std           9.908290
min          20.000000
25%          29.000000
50%          33.000000
75%          40.000000
max         143.000000
Name: member_age, dtype: float64

From the description above, the maximum age is 143, which is abnormal, I will drop ages which exceeds 100.

In [37]:
# Checking for ages abover 100 years

df.query('member_age > 100')

start_time                end_time start_station_id  \
1285   2019-02-28 19:29:17.627 2019-02-28 19:31:45.967            158.0   
10827  2019-02-27 19:21:34.436 2019-02-27 19:43:30.008            343.0   
16087  2019-02-27 08:37:36.864 2019-02-27 08:56:28.022            375.0   
17981  2019-02-26 19:53:38.528 2019-02-26 19:59:44.968             50.0   
18885  2019-02-26 17:48:59.016 2019-02-26 17:56:11.370             21.0   
...                        ...                     ...              ...   
171996 2019-02-03 17:33:54.607 2019-02-03 17:56:42.949             37.0   
173711 2019-02-03 09:45:30.464 2019-02-03 10:02:04.169            375.0   
177708 2019-02-01 19:09:28.387 2019-02-01 19:34:55.963            343.0   
177885 2019-02-01 18:38:40.471 2019-02-01 18:47:18.392             25.0   
182830 2019-02-01 07:45:05.934 2019-02-01 07:52:14.922            284.0   

                                       start_station_name  \
1285                        Shattuck Ave at Telegraph Ave   
10827                                 Bryant St at 2nd St   
16087                             Grove St at Masonic Ave   
17981                               2nd St at Townsend St   
18885    Montgomery St BART Station (Market St at 2nd St)   
...                                                   ...   
171996                                2nd St at Folsom St   
173711                            Grove St at Masonic Ave   
177708                                Bryant St at 2nd St   
177885                                Howard St at 2nd St   
182830  Yerba Buena Center for the Arts (Howard St at ...   

        start_station_latitude  start_station_longitude end_station_id  \
1285                 37.833279              -122.263490          173.0   
10827                37.783172              -122.393572          375.0   
16087                37.774836              -122.446546           36.0   
17981                37.780526              -122.390288           21.0   
18885                37.789625              -122.400811           50.0   
...                        ...                      ...            ...   
171996               37.785000              -122.395936          375.0   
173711               37.774836              -122.446546           36.0   
177708               37.783172              -122.393572          375.0   
177885               37.787522              -122.397405           30.0   
182830               37.784872              -122.400876           67.0   

                                         end_station_name  \
1285                              Shattuck Ave at 55th St   
10827                             Grove St at Masonic Ave   
16087                                 Folsom St at 3rd St   
17981    Montgomery St BART Station (Market St at 2nd St)   
18885                               2nd St at Townsend St   
...                                                   ...   
171996                            Grove St at Masonic Ave   
173711                                Folsom St at 3rd St   
177708                            Grove St at Masonic Ave   
177885     San Francisco Caltrain (Townsend St at 4th St)   
182830  San Francisco Caltrain Station 2  (Townsend St...   

        end_station_latitude  end_station_longitude bike_id   user_type  \
1285               37.840364            -122.264488    5391  Subscriber   
10827              37.774836            -122.446546    6249  Subscriber   
16087              37.783830            -122.398870    4968  Subscriber   
17981              37.789625            -122.400811    5226  Subscriber   
18885              37.780526            -122.390288    5319  Subscriber   
...                      ...                    ...     ...         ...   
171996             37.774836            -122.446546    4988  Subscriber   
173711             37.783830            -122.398870    5445  Subscriber   
177708             37.774836            -122.446546    5286  Subscriber   
177885        

## (Visualization 1)

> Write a comment about your visualization here. The visualization should be in the next cell, as a sub-slide type. Make sure your visualizations are polished!

> **Rubric Tip**: Provide at least 3 **polished** visualizations to convey key insights. The total number of visualizations in the slideshow shoould be less than 50% of the number of visualizations in the exploratory analysis. For example, if the exploratory analysis (Part I) has 18 visualizations, the slideshow can have (3 - 8) visualizations. 


> **Rubric Tip**: Each visualization in the slideshow is associated with **descriptive comments** that accurately depict their purpose and your observation. 


> **Rubric Tip**: All plots in the slideshow are appropriate, meaning the plot type, encodings, and transformations are suitable to the underlying data. 

> **Rubric Tip**: All plots in the slideshow are polished, meaning all plots have a title, labeled x/y axes (with units), x/y ticks, and legends. 

I willl start off with univariate distribution and visualize the relationship between variables.

## (Visualization 2)

> You should have at least three visualizations in your presentation,
but feel free to add more if you'd like!

## (Visualization 3)



### Generate Slideshow
Once you're ready to generate your slideshow, use the `jupyter nbconvert` command to generate the HTML slide show.  

In [ ]:
# Use this command if you are running this file in local
!jupyter nbconvert <Part_II_Filename>.ipynb --to slides --post serve --no-input --no-prompt

/bin/bash: line 1: Part_II_Filename: No such file or directory


> In the classroom workspace, the generated HTML slideshow will be placed in the home folder. 

> In local machines, the command above should open a tab in your web browser where you can scroll through your presentation. Sub-slides can be accessed by pressing 'down' when viewing its parent slide. Make sure you remove all of the quote-formatted guide notes like this one before you finish your presentation! At last, you can stop the Kernel. 

### Submission
If you are using classroom workspace, you can choose from the following two ways of submission:

1. **Submit from the workspace**. Make sure you have removed the example project from the /home/workspace directory. You must submit the following files:
   - Part_I_notebook.ipynb
   - Part_I_notebook.html or pdf
   - Part_II_notebook.ipynb
   - Part_I_slides.html
   - README.md
   - dataset (optional)


2. **Submit a zip file on the last page of this project lesson**. In this case, open the Jupyter terminal and run the command below to generate a ZIP file. 
```bash
zip -r my_project.zip .
```
The command abobve will ZIP every file present in your /home/workspace directory. Next, you can download the zip to your local, and follow the instructions on the last page of this project lesson.
